lib/swoosh/api_client/req.ex

```diff
@@ -0,0 +1,66 @@
+defmodule Swoosh.ApiClient.Req do
+  @moduledoc """
+  Req-based ApiClient for Swoosh.
+
+      config :swoosh, :api_client, Swoosh.ApiClient.Req
+
+  Any `client_options` that are set will be passed along to `Req.post` but the
+  following keys will be overwritten as they are set by Swoosh explicitly:
+    * `headers`
+    * `body`
+    * `decode_body` - set to false as Adapters expect the raw response
+  """
+
+  require Logger
+
+  @behaviour Swoosh.ApiClient
+  @user_agent {"User-Agent", "swoosh/#{Swoosh.version()}"}
+
+  @impl true
+  def init do
+    unless Code.ensure_loaded?(Req) do
+      Logger.error("""
+      Could not find req dependency.
+
+      Please add :req to your dependencies:
+
+          {:req, "~> 0.4"}
+
+      Or set your own Swoosh.ApiClient:
+
+          config :swoosh, :api_client, MyAPIClient
+      """)
+
+      raise "missing req dependency"
+    end
+
+    _ = Application.ensure_all_started(:req)
+    :ok
+  end
+
+  @impl true
+  def post(url, headers, body, %Swoosh.Email{} = email) do
+    url = IO.iodata_to_binary(url)
+
+    options = email.private[:client_options] || []
+
+    # Set `decode_body` option to false as Swoosh's Mailgun adapter expects
+    # that the body is a binary. Req will run `Jason.decode` itself and
+    # `response.body` will already be a map if it is not set.
+    required_options = [
+      headers: [@user_agent | headers],
+      body: body,
+      decode_body: false
```

Comment:
Blast from the past! Similarly to `decode_body: false`, you might consider setting `retry: false`. Req is retrying by default, though I'm considering changing that before v1.0, slightly leaning towards. And so it wouldn't hurt to be super explicit here.

In swoosh README we have:

> One of the downsides of sending email asynchronously is that failures won't be reported to the user, who won't have an opportunity to try again immediately, and tasks by default do not retry on errors.

So yeah this is not totally accurate with regards to Req right and it can be indeed surprising to folks, especially if they end up putting stuff to Oban.

Apologies I haven't mentioned that sooner. Thanks for adding built-in support for Req.

Reply:
Shouldn't we have `retry: true` instead? :) I guess people can configure it as necessary too, since options are merged below, so perhaps no changes necessary here.

Reply:
Doing `retry: true` here in swoosh would be super explicit and protect against any changes in req in this area if that’s what you meant.

Reply:
Sorry, the current default is retry: :safe_transient. Safe meaning only GET/HEAD. So this is moot, I assume sending emails would be using POST. Sorry for noise!